### PR TITLE
components: Add background color utilites

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Description
Adds two new utilities for deriving background and text color from a small set of supported colors. This is used by the Tag and Badge components to determine what colors to use for each variation.

## How has this been tested?
Test this in Storybook is the best way because we want to be able to test that the contrast ratios are enough for each color. I had to make some tweaks for three of the colors from G2 because they did not have sufficient color contrasts. Using the Axe browser extension or the built in accessibility tools in Firefox, check each rendered example in the storybook story here: `/iframe.html?id=components-experimental-background-colors--default&args=` (go to that URL after running `npm run storybook:dev`)

Ensure that each example has the sufficient color contrast to have at least 4.5 (i.e., AA) contrast.

## Screenshots 
Captura de Tela 2021-07-12 às 09 04 20

## Types of changes
New features.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by sarayourfriend** [comment]
As part of the story for https://github.com/WordPress/gutenberg/pull/33430 I noticed that the bold variations need some attention for the contrast to be correct. I'll fix that up in this PR.

**Comment by diegohaz** [comment]
Because it's using alpha, the contrast ratio will change based on the background:

screenshot of storybook with dark background showing bad contrasts

It's an extreme example, but this makes the result unpredictable anyway.

Can we make the colors opaque and just change their lightness or is it a requirement to use alpha?

**Comment by sarayourfriend** [comment]
> Can we make the colors opaque and just change their lightness or is it a requirement to use alpha?

Not a requirement to use alpha, we can definitely just lighten them. Good catch!

**Comment by sarayourfriend** [comment]
@diegohaz Okay it should be fixed now! The logic is definitely a little more cumbersome, if you can think of any ways to improve please chime in 🙇

File: packages/components/src/utils/backgrounds.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+	if ( isBold ) {
+		colorized.darken( 20 );
+	}
+
+	// if ( isBold ) {
+	// 	colorized.lighten( 30 );
+	// } else if ( isSubtle ) {
+	// 	colorized.lighten( 90 );
+	// }
```
Comment (ciampo) [inline packages/components/src/utils/backgrounds.ts]: Can this code be removed? Is the `isSubtle` option still a requirement, or can we do without it?

Comment (sarayourfriend) [reply packages/components/src/utils/backgrounds.ts]: isSubtle is complete unused so I removed it. The comments left were a mistake. Removed in the next commit

**Comment by ciampo** [comment]
Do we need to get some sort of design review regarding the colors used? In particular about the amount of lightening/darkening/desaturation and how we darken the background color when `isBold` is specified?

File: packages/components/src/utils/backgrounds.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+	'red',
+	'purple',
+	'green',
+	'yellow',
+	'orange',
+	'darkGray',
+	'lightGray',
+] as const;
+
+export type SupportedColors = typeof SUPPORTED_COLORS[ number ];
```
Comment (ciampo) [inline packages/components/src/utils/backgrounds.ts]: Nit: should this type be named `SupportedColor` (singular) ?

**Comment by sarayourfriend** [comment]
I have no idea who to ask for design reviews.

**Pull request closed without merging**